app/domain/authentication/authn_oidc/v2/client.rb

```diff
@@ -37,11 +37,31 @@ def oidc_client
           end
         end
 
-        def callback(code:, nonce:, code_verifier:)
+        def get_token_with_code(code:, nonce:, code_verifier:)
           oidc_client.authorization_code = code
+          id_token, refresh_token = get_token_pair(code_verifier)
+          decoded_id_token = decode_id_token(id_token)
+          verify_id_token(decoded_id_token, nonce)
+
+          [decoded_id_token, refresh_token]
+        end
+
+        def get_token_with_refresh_token(refresh_token:, nonce:)
+          oidc_client.refresh_token = refresh_token
+          id_token, refresh_token = get_token_pair(nil)
+          decoded_id_token = decode_id_token(id_token)
+
+          nonce_to_verify = nonce
+          nonce_to_verify = nil unless decoded_id_token.raw_attributes['nonce']
```

Comment:
Is the idea here that returning the original `nonce` value is optional for the authorization server when using the refresh_token grant type, versus required when making the original authorization code request?

I'm wondering if this could be handled more cleanly in the `verify_id_token` method

Reply:
Exactly - more detail [here](https://bitbucket.org/openid/connect/issues/1025/ambiguity-with-how-nonce-is-handled-on):

> it SHOULD NOT have a `nonce` Claim, even when the ID Token issued at the time of the original authentication contained `nonce`; however, if it is present, its value MUST be the same as in the ID Token issued at the time of the original authentication

> I'm wondering if this could be handled more cleanly in the verify_id_token method

You're probably right - I'll refactor.